src/Http/Middleware/SessionCsrfProtectionMiddleware.php

```diff
@@ -0,0 +1,217 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http\Middleware;
+
+use ArrayAccess;
+use Cake\Http\Exception\InvalidCsrfTokenException;
+use Cake\Http\Session;
+use Cake\Utility\Hash;
+use Cake\Utility\Security;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use RuntimeException;
+
+/**
+ * Provides CSRF protection via session based tokens.
+ *
+ * This middleware adds a CSRF token to the session. Each request must
+ * contain a token in request data, or the X-CSRF-Token header on each PATCH, POST,
+ * PUT, or DELETE request. This follows a 'synchronizer token' pattern.
+ *
+ * If the request data is missing or does not match the session data,
+ * an InvalidCsrfTokenException will be raised.
+ *
+ * This middleware integrates with the FormHelper automatically and when
+ * used together your forms will have CSRF tokens automatically added
+ * when `$this->Form->create(...)` is used in a view.
+ *
+ * If you use this middleware *do not* also use CsrfProtectionMiddleware.
+ *
+ * @see https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#sychronizer-token-pattern
+ */
+class SessionCsrfProtectionMiddleware implements MiddlewareInterface
+{
+    /**
+     * Config for the CSRF handling.
+     *
+     *  - `key` The session key to use. Defaults to `csrfToken`
+     *  - `field` The form field to check. Changing this will also require configuring
+     *    FormHelper.
+     *
+     * @var array
+     */
+    protected $_config = [
```

Comment:
```suggestion
    protected $config = [
```
What is the cakephp policy for underscore protected? Are we transitioning away? It seems we randomly avoid them and randomly copy them.

Reply:
We generally avoid the underscore for new protected properties but for `$_config` I would prefer keeping it for consistency with other classes.

We'll have to live with this till v5.

Reply:
We could start making things better now so there are fewer things to change in the future.